pkg/rest/core/fake/rest_client.go

```diff
@@ -188,13 +194,110 @@ func getRouter(storage NamespacedStorage) http.Handler {
 		deleteItem(storage),
 	).Methods("DELETE")
 	r.HandleFunc(
+		"/apis/servicecatalog.k8s.io/v1alpha1/watch/namespaces/{namespace}/{type}/{name}",
+		watchItem(watcher),
+	).Methods("GET")
+	r.HandleFunc(
+		"/apis/servicecatalog.k8s.io/v1alpha1/watch/namespaces/{namespace}/{type}",
+		watchList(watcher),
+	).Methods("GET")
+	r.HandleFunc(
 		"/api/v1/namespaces",
 		listNamespaces(storage),
-	)
+	).Methods("GET")
 	r.NotFoundHandler = http.HandlerFunc(notFoundHandler)
 	return r
 }
 
+func watchItem(watcher *Watcher) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ch := watcher.ReceiveChan()
+		for evt := range ch {
+			codec, err := testapi.GetCodecForObject(evt.Object)
+			if err != nil {
+				errStr := fmt.Sprintf("error getting codec (%s)", err)
+				log.Fatal(errStr)
+				http.Error(w, errStr, http.StatusInternalServerError)
+				return
+			}
+			objBytes, err := runtime.Encode(codec, evt.Object)
+			if err != nil {
+				errStr := fmt.Sprintf("error encoding item (%s)", err)
+				log.Fatal(errStr)
+				http.Error(w, errStr, http.StatusInternalServerError)
+				return
+			}
+
+			evt := metav1.WatchEvent{
+				Type: fmt.Sprintf("%s", evt.Type),
+				Object: runtime.RawExtension{
+					Object: evt.Object,
+					Raw:    objBytes,
+				},
+			}
+			b, err := json.Marshal(&evt)
+			if err != nil {
+				errStr := fmt.Sprintf("error encoding JSON (%s)", err)
+				log.Fatal(errStr)
+				http.Error(w, errStr, http.StatusInternalServerError)
+				return
+			}
+			w.Header().Set("Content-Type", "application/json, */*")
+			w.Write(b)
+		}
+	}
+}
+
+func watchList(watcher *Watcher) func(http.ResponseWriter, *http.Request) {
+	const timeout = 1 * time.Second
+	return func(w http.ResponseWriter, r *http.Request) {
+		ch := watcher.ReceiveChan()
+		for {
+			select {
+			case <-time.After(timeout):
+				errStr := fmt.Sprintf("didn't receive within %s", timeout)
+				log.Fatal(errStr)
+				http.Error(w, errStr, http.StatusInternalServerError)
+			case evt, ok := <-ch:
+				if !ok {
+					return
+				}
+				codec, err := testapi.GetCodecForObject(evt.Object)
```

Comment:
There are a couple opportunities to extract methods here and in `watch` you could do in follow-ups.